prospero-common/src/main/java/org/wildfly/prospero/galleon/ChannelMavenArtifactRepositoryManager.java

```diff
@@ -52,40 +64,155 @@ public ChannelMavenArtifactRepositoryManager(ChannelSession channelSession, Chan
+
+        if (found.isPresent()) {
+            result = channelSession.resolveDirectMavenArtifact(artifact.getGroupId(), artifact.getArtifactId(), artifact.getExtension(),
+                    artifact.getClassifier(), found.get().getVersion());
+        } else if (isUniverseOrProducerArtifact(artifact.getArtifactId())) {
+            result = channelSession.resolveMavenArtifact(artifact.getGroupId(), artifact.getArtifactId(), artifact.getExtension(),
+                    artifact.getClassifier(), null);
+        } else {
+            throw new MavenUniverseException(ProsperoLogger.ROOT_LOGGER.unableToResolve() + " ["+ artifact.getCoordsAsString()+"]");
+        }
+        return result;
+    }
+
+    private boolean requiresChannel(MavenArtifact artifact) {
+        boolean requireChannel = Boolean.parseBoolean(artifact.getMetadata().get(REQUIRE_CHANNEL_FOR_ALL_ARTIFACT));
+        try {
+            if (!requireChannel && ! fpRequireChannel(artifact)) {
+                return false;
+            } else {
+                return true;
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
         }
     }
 
     @Override
     public void resolveAll(Collection<MavenArtifact> artifacts) throws MavenUniverseException {
-        final MavenArtifactMapper mapper = new MavenArtifactMapper(artifacts);
-        if (manifest == null) {
-            final List<org.wildfly.channel.MavenArtifact> channelArtifacts = channelSession.resolveMavenArtifacts(mapper.toChannelArtifacts());
 
+        if (manifest == null) {
+            // split the artifacts into requiring channels and not requiring channels
+            final List<MavenArtifact> artifactsRequiringChannels = artifacts.stream()
+                    .filter(a -> requiresChannel(a))
+                    .collect(Collectors.toList());
+            final List<MavenArtifact> artifactsNotRequiringChannels = artifacts.stream()
+                    .filter(a -> !requiresChannel(a))
+                    .collect(Collectors.toList());
+            // bulk resolve artifacts requiring channels - if any fail, throw exception
+            MavenArtifactMapper mapper = new MavenArtifactMapper(artifactsRequiringChannels);
+            List<org.wildfly.channel.MavenArtifact> channelArtifacts = channelSession.resolveMavenArtifacts(mapper.toChannelArtifacts());
             mapper.applyResolution(channelArtifacts);
-        } else {
 
+            // bulk resolve other artifacts
+            final MavenArtifactMapper mapperNotRequiringChannels = new MavenArtifactMapper(artifactsNotRequiringChannels);
+            try {
+                channelArtifacts = channelSession.resolveMavenArtifacts(mapperNotRequiringChannels.toChannelArtifacts());
+                mapperNotRequiringChannels.applyResolution(channelArtifacts);
+            } catch (UnresolvedMavenArtifactException e) {
+                if (e.getCause() instanceof ArtifactResolutionException) {
```

Comment:
The cause looks like won't be ArtifactResolutoinException from https://github.com/wildfly-extras/wildfly-channel/blob/1.0.0.Beta6/core/src/main/java/org/wildfly/channel/ChannelSession.java#L249. The cause is the UnresolvedMavenArtifactException is itself in my test.

Reply:
@gaol do you have a stack trace for that exception by any chance?

Reply:
here you are:
```java
Downloading artifactsorg.wildfly.channel.UnresolvedMavenArtifactException: Can not resolve latest Maven artifact (no stream found) : org.wildfly:wildfly-web:jar:
	at org.wildfly.channel.ChannelSession.lambda$findChannelWithLatestVersion$6(ChannelSession.java:251)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at org.wildfly.channel.ChannelSession.findChannelWithLatestVersion(ChannelSession.java:243)
	at org.wildfly.channel.ChannelSession.splitArtifactsPerChannel(ChannelSession.java:258)
	at org.wildfly.channel.ChannelSession.resolveMavenArtifacts(ChannelSession.java:118)
	at org.wildfly.prospero.galleon.ChannelMavenArtifactRepositoryManager.resolveAll(ChannelMavenArtifactRepositoryManager.java:183)
	at org.wildfly.galleon.plugin.WfInstallPlugin.resolveArtifactsInCache(WfInstallPlugin.java:516)
	at org.wildfly.galleon.plugin.WfInstallPlugin.postInstall(WfInstallPlugin.java:375)
	at org.jboss.galleon.runtime.ProvisioningRuntime$3.visitPlugin(ProvisioningRuntime.java:269)
	at org.jboss.galleon.runtime.ProvisioningRuntime$3.visitPlugin(ProvisioningRuntime.java:266)
	at org.jboss.galleon.layout.ProvisioningLayout$Handle.visitPlugins(ProvisioningLayout.java:287)
	at org.jboss.galleon.layout.ProvisioningLayout.visitPlugins(ProvisioningLayout.java:860)
	at org.jboss.galleon.runtime.ProvisioningRuntime.provision(ProvisioningRuntime.java:266)
	at org.jboss.galleon.ProvisioningManager.doProvision(ProvisioningManager.java:646)
	at org.jboss.galleon.ProvisioningManager.provision(ProvisioningManager.java:405)
	at org.wildfly.prospero.cli.commands.UpdateCommand$GenerateCommand.lambda$call$0(UpdateCommand.java:331)
	at org.wildfly.prospero.galleon.GalleonUtils.executeGalleon(GalleonUtils.java:85)
	at org.wildfly.prospero.cli.commands.UpdateCommand$GenerateCommand.call(UpdateCommand.java:331)
	at org.wildfly.prospero.cli.commands.UpdateCommand$GenerateCommand.call(UpdateCommand.java:291)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at org.wildfly.prospero.cli.CliMain.main(CliMain.java:55)

```

Reply:
Ah that's when the stream can't be found in first place. Good catch, thanks @gaol!